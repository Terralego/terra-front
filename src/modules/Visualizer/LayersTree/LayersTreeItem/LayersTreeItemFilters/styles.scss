@use "sass:math";

@import '../../../../../utils/variables.scss';

.layersTreeItemFilter {
  margin: .5rem -0.2rem 0.15rem;
  &__tag {
    margin: 0.2rem;
    max-width: calc(100% - 1rem);
    .bp3-text-overflow-ellipsis {
      white-space: inherit;
    }
  }
}

@media(min-width: $breakpoints-phone) and (max-width: $breakpoints-desktop){
  .layersTreeItemFilter {
    margin: 0;
    &__tag {
      padding: .25rem .25rem .25rem .5rem;
      margin: 1rem 0 0 0;
      max-width: calc(100% - 1rem / #{$scaleRatio});
      .bp3-text-overflow-ellipsis {
        white-space: inherit;
        font-size: math.div(.75rem, $scaleRatio);
      }
      .bp3-tag-remove {
        padding: .5rem;
      }
    }
  }
}
