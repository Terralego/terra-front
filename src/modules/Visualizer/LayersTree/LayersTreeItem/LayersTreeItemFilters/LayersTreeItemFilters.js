import React from 'react';
import { Tag } from '@blueprintjs/core';

import './styles.scss';

function getLiteralValue (value) {
  if (value instanceof Date) {
    return `le ${value.toLocaleDateString()}`;
  }
  return value;
}

export function getValueFromType (value, type) {
  if (type === 'boolean') {
    return null;
  }
  if (type === 'range') {
    return `: Entre ${getLiteralValue(value[0])} et ${getLiteralValue(value[1])}`;
  }
  return `: ${getLiteralValue(value)}`;
}

const LayersTreeItemFilters = ({
  filtersValues,
  setLayerState,
  layer,
  layer: { filters: { form: layerForm } = {} } = {},
  activeLayer: { filters: { form = layerForm } = {} } = {},
}) => {
  const properties = (form || []).map(({ property, ...rest }) => ({
    ...rest,
    property,
    value: filtersValues[property],
  }));

  if (!properties.some(({ value }) => value)) {
    return null;
  }
  return (
    <div className="layersTreeItemFilter">
      {properties.map(({ property, value, type, label }) => (
        (value && value.length > 0) && (
        <Tag
          key={property}
          className="layersTreeItemFilter__tag"
          onRemove={() =>
            setLayerState({
              layer,
              state: {
                filters: { ...filtersValues, [property]: undefined },
                total: undefined,
              },
            })}
        >
          {label}
          {getValueFromType(value, type)}
        </Tag>
        )
      ))}
    </div>
  );
};

export default LayersTreeItemFilters;
