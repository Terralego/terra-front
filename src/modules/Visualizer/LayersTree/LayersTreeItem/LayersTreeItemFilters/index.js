import LayersTreeItemFilters from './LayersTreeItemFilters';
import { connectLayersTree } from '../../LayersTreeProvider/context';

export default connectLayersTree(({
  getLayerState, setLayerState, translate,
}, {
  layer, layer: { exclusive, layers },
}) => {
  const activeLayer = exclusive
    ? layers.find(l => getLayerState({ layer: l }).active) || layer
    : layer;

  return {
    filtersValues: getLayerState({ layer: activeLayer }).filters || {},
    setLayerState,
    activeLayer,
    translate,
  };
})(LayersTreeItemFilters);
