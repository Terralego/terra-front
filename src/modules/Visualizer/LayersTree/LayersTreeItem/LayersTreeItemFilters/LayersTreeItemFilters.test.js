import React from 'react';
import renderer from 'react-test-renderer';

import LayersTreeItemFilters, { getValueFromType } from './LayersTreeItemFilters';

let toLocaleDateString;
beforeEach(() => {
  // eslint-disable-next-line prefer-destructuring
  toLocaleDateString = Date.prototype.toLocaleDateString;
  // eslint-disable-next-line no-extend-native
  Date.prototype.toLocaleDateString = () => 'mocked date';
});
afterEach(() => {
  // eslint-disable-next-line no-extend-native
  Date.prototype.toLocaleDateString = toLocaleDateString;
});

it('should render', () => {
  const tree = renderer.create(
    <>
      <LayersTreeItemFilters
        filtersValues={{
          single: 'foo',
          many: ['foo', 'bar'],
          range: [2, 42],
          date: [new Date(), new Date(+new Date() + 1000000)],
          bool: true,
        }}
        layer={{
          filters: {
            form: [{
              type: 'single',
              property: 'single',
            }, {
              type: 'many',
              property: 'many',
              values: ['foo', 'bar', 'lorem'],
            }, {
              type: 'range',
              property: 'range',
            }, {
              type: 'range',
              property: 'date',
              format: 'date',
            }, {
              type: 'boolean',
              property: 'bool',
            }],
          },
        }}
      />
      <LayersTreeItemFilters
        activeLayer={{}}
      />
    </>,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should remove tag', () => {
  const layer = {
    filters: {
      form: [{
        type: 'single',
        property: 'single',
      }],
    },
  };
  const setLayerState = jest.fn();
  const tree = renderer.create(
    <LayersTreeItemFilters
      setLayerState={setLayerState}
      filtersValues={{
        single: 'foo',
      }}
      layer={layer}
    />,
  );
  const tag = tree.root.find(({ type: { displayName } }) => displayName === 'Blueprint3.Tag');
  tag.props.onRemove();
  expect(setLayerState).toHaveBeenCalledWith({
    layer,
    state: {
      filters: {},
      total: undefined,
    },
  });
});

it('should return no value', () => {
  getValueFromType(true, 'boolean');
  expect(getValueFromType(true, 'boolean')).toEqual(null);
});
