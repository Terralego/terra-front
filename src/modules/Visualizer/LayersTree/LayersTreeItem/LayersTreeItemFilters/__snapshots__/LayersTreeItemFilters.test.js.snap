// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<div
  className="layersTreeItemFilter"
>
  <span
    className="bp3-tag layersTreeItemFilter__tag"
  >
    <span
      className="bp3-text-overflow-ellipsis bp3-fill"
    >
      foo
    </span>
    <button
      className="bp3-tag-remove"
      onClick={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-small-cross"
        icon="small-cross"
      >
        <svg
          data-icon="small-cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            small-cross
          </desc>
          <path
            d="M9.41 8l2.29-2.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42L6.59 8 4.3 10.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l2.29 2.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </span>
  <span
    className="bp3-tag layersTreeItemFilter__tag"
  >
    <span
      className="bp3-text-overflow-ellipsis bp3-fill"
    >
      foo, bar
    </span>
    <button
      className="bp3-tag-remove"
      onClick={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-small-cross"
        icon="small-cross"
      >
        <svg
          data-icon="small-cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            small-cross
          </desc>
          <path
            d="M9.41 8l2.29-2.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42L6.59 8 4.3 10.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l2.29 2.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </span>
  <span
    className="bp3-tag layersTreeItemFilter__tag"
  >
    <span
      className="bp3-text-overflow-ellipsis bp3-fill"
    >
      Range of foo: 
      Between 2 and 42
    </span>
    <button
      className="bp3-tag-remove"
      onClick={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-small-cross"
        icon="small-cross"
      >
        <svg
          data-icon="small-cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            small-cross
          </desc>
          <path
            d="M9.41 8l2.29-2.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42L6.59 8 4.3 10.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l2.29 2.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </span>
  <span
    className="bp3-tag layersTreeItemFilter__tag"
  >
    <span
      className="bp3-text-overflow-ellipsis bp3-fill"
    >
      Range of date: 
      Starting from mocked date to mocked date
    </span>
    <button
      className="bp3-tag-remove"
      onClick={[Function]}
      type="button"
    >
      <span
        className="bp3-icon bp3-icon-small-cross"
        icon="small-cross"
      >
        <svg
          data-icon="small-cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
        >
          <desc>
            small-cross
          </desc>
          <path
            d="M9.41 8l2.29-2.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-2.29-2.3a1.003 1.003 0 00-1.42 1.42L6.59 8 4.3 10.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l2.29 2.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
            fillRule="evenodd"
          />
        </svg>
      </span>
    </button>
  </span>
</div>
`;
